plugins/catalog-backend/src/next/Stitcher.ts

```diff
@@ -49,64 +51,123 @@ export class Stitcher {
+          .from('refresh_state')
+          .leftJoin('incoming_references', {})
+          .leftOuterJoin('final_entities', {
+            'final_entities.entity_id': 'refresh_state.entity_id',
+          })
+          .leftOuterJoin('relations', {
+            'relations.source_entity_ref': 'refresh_state.entity_ref',
+          })
+          .where({ 'refresh_state.entity_ref': entityRef })
+          .orderBy('relationType', 'asc')
+          .orderBy('relationTarget', 'asc');
+
+        // If there were no rows returned, it would mean that there was no
+        // matching row even in the refresh_state. This can happen for example
+        // if we emit a relation to something that hasn't been ingested yet.
+        // It's safe to ignore this stitch attempt in that case.
+        if (!result.length) {
           this.logger.debug(
             `Unable to stitch ${entityRef}, item does not exist in refresh state table`,
           );
           return;
-        } else if (!result.processed_entity) {
+        }
+
+        const {
+          entityId,
+          processedEntity,
+          // errors,
+          incomingReferenceCount,
+          previousEtag,
+        } = result[0];
+
+        // If there was no processed entity in place, the target hasn't been
+        // through the processing steps yet. It's safe to ignore this stitch
+        // attempt in that case, since another stitch will be triggered when
+        // that processing has finished.
+        if (!processedEntity) {
           this.logger.debug(
             `Unable to stitch ${entityRef}, the entity has not yet been processed`,
           );
           return;
         }
 
-        const entity: Entity = JSON.parse(result.processed_entity);
-
-        const entityId = entity?.metadata?.uid;
-        if (!entityId) {
-          this.logger.error(`missing ID in entity ${JSON.stringify(entity)}`);
-          return;
-        }
+        // Grab the processed entity and stitch all of the relevant data into it
+        const entity = JSON.parse(processedEntity) as Entity;
+        const isOrphan = Number(incomingReferenceCount) === 0;
 
-        const [reference_count_result] = await tx<DbRefreshStateReferences>(
-          'refresh_state_references',
-        )
-          .where({ target_entity_id: entity.metadata.uid })
-          .count({ reference_count: 'target_entity_id' });
-
-        if (Number(reference_count_result.reference_count) === 0) {
-          this.logger.debug(`${entityRef} is orphan`);
+        if (isOrphan) {
+          this.logger.debug(`${entityRef} is an orphan`);
           entity.metadata.annotations = {
             ...entity.metadata.annotations,
             ['backstage.io/orphan']: 'true',
           };
         }
 
-        const relationResults = await tx<DbRelationsRow>('relations')
-          .where({ source_entity_ref: entityRef })
-          .select();
+        // TODO: entityRef is lower case and should be uppercase in the final result
```

Comment:
Should it? O_o

Reply:
yeah i think it should - the output is a structure, and it just looks bad to get back `"target": { "kind": "component" ...` in the json for no good reason other than us cheaping out and not storing the actual original values :)